pipeliner.go

```diff
@@ -0,0 +1,246 @@
+		return !blockingCmds[strings.ToUpper(v.cmd)]
+	case pipeline:
+		// do not merge user defined pipelines with our own pipelines so that
+		// users can better control pipelining
+		return false
+	case pipedCmdPipeline:
+		// prevent accidental cycles / loops by disallowing pipelining of
+		// pipes created by a pipeliner.
+		return false
+	case CmdAction:
+		// there is currently no way to get the command for CmdAction implementations
+		// from outside the radix package so we can not multiplex these commands.
+		return false
+	default:
+		return false
+	}
+}
+
+// Do executes the given Action as part of the pipeline.
+//
+// If a is not a CmdAction, Do panics.
+func (p *pipeliner) Do(a Action) error {
+	req := getPipedCmd(a.(CmdAction)) // get this outside the lock to avoid
+
+	p.l.RLock()
+	if p.closed {
+		p.l.RUnlock()
+		return errClientClosed
+	}
+	p.reqCh <- *req
+	p.l.RUnlock()
+
+	err := <-req.resCh
+	poolPipedCmd(req)
+	return err
+}
+
+// Close closes the pipeliner and makes sure that all background goroutines
+// are stopped before returning.
+//
+// Close does *not* close the underyling Client.
+func (p *pipeliner) Close() error {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	if p.closed {
+		return nil
+	}
+
+	close(p.reqCh)
+	p.reqWG.Wait()
+
+	for i := 0; i < cap(p.flushSema); i++ {
+		<-p.flushSema
+	}
+
+	p.c = nil
+	return nil
+}
+
+func (p *pipeliner) reqLoop() {
+	var pipe []pipedCmd
+	if p.limit > 0 {
+		pipe = make(pipedCmdPipeline, 0, p.limit)
+	}
+
+	t := getTimer(time.Hour)
+	defer putTimer(t)
+
+	t.Stop()
+
+	for {
+		select {
+		case req, ok := <-p.reqCh:
+			if !ok {
+				p.flush(pipe)
+				return
+			}
+
+			pipe = append(pipe, req)
+
+			if p.limit > 0 && len(pipe) == p.limit {
+				// if we reached the pipeline limit, execute now to avoid unnecessary waiting
+				t.Stop()
+
+				p.flush(pipe)
+				pipe = pipe[:0]
+			} else if len(pipe) == 1 {
+				t.Reset(p.window)
+			}
+		case <-t.C:
+			p.flush(pipe)
+			pipe = pipe[:0]
+		}
+	}
+}
+
+func (p *pipeliner) flush(pipe []pipedCmd) {
+	if len(pipe) == 0 {
+		return
+	}
+
+	// copy requests into a pipeline so that we can flush the pipeline in
+	// the background, to avoid blocking the main loop.
+	pipeCopy := make(pipedCmdPipeline, len(pipe))
+	copy(pipeCopy, pipe)
+
+	<-p.flushSema
+	go func() {
+		defer func() {
+			p.flushSema <- struct{}{}
+		}()
+
+		if err := p.c.Do(pipeCopy); err != nil {
+			for _, req := range pipeCopy {
+				req.resCh <- err
+			}
+		}
+	}()
+}
+
+type pipedCmd struct {
+	cmd   CmdAction
+	resCh chan error
+}
+
+var pipedCmdPool sync.Pool
+
+func getPipedCmd(cmd CmdAction) *pipedCmd {
+	req, _ := pipedCmdPool.Get().(*pipedCmd)
+	if req != nil {
+		req.cmd = cmd
+		return req
+	}
+	return &pipedCmd{
+		cmd: cmd,
+		// using a buffer of 1 is faster than no buffer in most cases
+		resCh: make(chan error, 1),
+	}
+}
+
+func poolPipedCmd(req *pipedCmd) {
+	req.cmd = nil
+	pipedCmdPool.Put(req)
+}
+
+type pipedCmdPipeline []pipedCmd
```

Comment:
Done, although it was a bit more complex than this. I had to embed `pipeline` into the `pipelinerPipeline` and change the code to work with pointers to `pipelinerCmd` to avoid allocations (since adding the `pipelinerCmd` struct to a `pipeline` involves boxing)